packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,158 @@
+
+### Lazy Loading
+
+In addition to just supporting data hierarchy with a fully provided set of data, we also need to support the scenario where a row is presented as a parent, but none of its children are currently represented in the data. Upon expanding the row, the client is then expected to load all of the data for that parent and provide it to the table (via the standard `setData` method).
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#890: Design for hierarchical data support in tables](https://github.com/ni/nimble/issues/890)
+-   [#861: Table hierarchical data support](https://github.com/ni/nimble/issues/861)
+-   [Prototype using flat list that includes a parentId](https://github.com/ni/nimble/tree/data-hierarchy-flat-list-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-qzwoshcidz.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+-   [Prototype with user data being hierarchical](https://github.com/ni/nimble/tree/data-hierarchy-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-yncupvnoes.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+
+## Implementation / Design
+
+Users will provide hierarchical data as a flat list (just as they do for non-hierarchical data), but will have to include a field in their data that is representative of the "parent Id" for that record of data.
+
+### `Table` API
+
+```ts
+public Table() {
+    ...
+    // This attribute determines which field in the record specifies the id that is the
+    // parent row. If it is not set, the row has no parent.
+    @attr({ attribute: 'parent-id-field-name' })
+    public parentIdFieldName?: string;
+
+    // This field would be an option for how we allow a user to specify the field
+    // name where they will provide a value indicating whether or not that row
+    // is a parent row. See the below discussion on lazy-loading. If it is not set,
+    // the row is presumed child-less (and thus will show no expand/collapse button).
+    @attr({ attribute: 'has-children-field-name' })
+    public hasChildrenFieldName?: string;
+
+    // This attribute will control the behavior of what happens when a user selects a parent
+    // row.
+    @attr({ attribute: 'parent-selects-children'})
+    public parentSelectsChildren?: boolean;
+
+    // The set of rows the user would like to toggle the expand/collapse state of
+    public async expandCollapseRows(rowIds: string[]): Promise<void> {
```

Comment:
Would this toggle the state of all of the provided `rowIds`, possibly resulting in some of them becoming expanded and others collapsed? If so, that seems somewhat messy to me because calling this requires knowing the state of every row first. I wonder if it might be more convenient to have one of these:
1. `expandRows()` and `collapseRows()` which don't toggle but just set the state on the rows you provide, leaving other rows unchanged
2. `collapseAllAndExpandRows()` where you provide the entire set of expanded rows and everything else ends up collapsed

I'm struggling to weigh the options here. Do you have ideas of what these would be used for?

Reply:
I agree that it seems like a questionable API for a "toggle row expanded" function when I would think a client would want to be explicitly expanding or collapsing (rather than toggling) a row. I also think "expand" might be an interesting case -- would we expand all ancestors and the row specified? Maybe that's rationale for just having a "collapse" function for now (although maybe it's too inconsistent to have collapse and not expand).

Reply:
The driving force behind this API was the need for a user to easily collapse a row (without explicit interaction) when they hit an error case when wanting to populate a parent row with children (they expanded a parent row, and now the 'row loading' indicator is present which they want to remove).

From the suggestions, I think I'm happiest with the `expandRows()` and `collapseRows()` APIs. Can we provide an additional parameter that controls whether the action should apply to all children:

```ts
public expandRows(rowIds: string[], expandChildren?: boolean): Promise<void>{}
```